next-frontend/src/lib/wca/users/usePermissions.ts

```diff
@@ -0,0 +1,46 @@
+import useAPI from "@/lib/wca/useAPI";
+import { useSession } from "next-auth/react";
+import { components } from "@/lib/wca/wcaSchema";
+import { useQuery } from "@tanstack/react-query";
+
+interface usePermissionsResponse {
+  permissions: components["schemas"]["UserPermissions"];
+  isLoading: boolean;
+}
+
+const NO_PERMISSIONS = {
+  can_access_panels: { scope: [] },
+  can_administer_competitions: { scope: [] },
+  can_attend_competitions: { scope: [] },
+  can_create_groups: { scope: [] },
+  can_edit_delegate_report: { scope: [] },
+  can_edit_groups: { scope: [] },
+  can_organize_competitions: { scope: [] },
+  can_read_groups_current: { scope: [] },
+  can_read_groups_past: { scope: [] },
+  can_request_to_edit_others_profile: { scope: [] },
+  can_view_delegate_admin_page: { scope: [] },
+  can_view_delegate_report: { scope: [] },
+}
+
+export default function usePermissions(): usePermissionsResponse {
+  const { data: session } = useSession();
+  const api = useAPI();
+  const { data: request, isLoading } = useQuery({
+    enabled: Boolean(session),
+    queryKey: ["permissions", session?.user.id],
+    queryFn: () => api.GET("/users/me/permissions")
+  })
+
+  if(!session){
+    return {
+      permissions: NO_PERMISSIONS,
+      isLoading: false,
+    };
+  }
+
+  return {
+    isLoading,
+    permissions: request?.data || NO_PERMISSIONS
+  }
```

Comment:
We wouldn't want to access permissions while they're loading right? I would just return `{ isLoading: true }` if it's loading, and `{ isLoading: false, permissions: request.data }` otherwise.

To avoid any awkwardness with having to check `isLoading` every time before accessing permissions, a wrapper can wait for loading to finish, and only render children after.

I did this in a work project and it worked well.

Reply:
how would such a wrapper look like? would that use the react 19 use hook? I saw something in the next doc, but I think it would require a paradigm shift for how we do requests.

Reply:
```
PermissionWrapper({ children }) {
  const permissions = usePermissions();
  if (permissions.isLoading) return <LoadingUI />
  return (
    <PermissionProvider value={permissions.value}>
      {children}
    </PermissionProvider>
  )
}
```
and then access them in the children via the context.

Apply at the router root (preferable), or a particular router branch (next best), or on each individual component that needs permission, depending on how things are set up.

I'm not familiar with `use` (but it's apparently _not_ a hook).